source/containerimage/pull.go

```diff
@@ -106,19 +106,13 @@ func (is *Source) ResolveImageConfig(ctx context.Context, ref string, opt llb.Re
 		}
 		rslvr = resolver.DefaultPool.GetResolver(is.RegistryHosts, ref, "pull", sm, g).WithImageStore(is.ImageStore, rm)
 	case ResolverTypeOCILayout:
-		// with OCI layout, we always just "pull"
 		rm = source.ResolveModeForcePull
-		// get the content store ID from the ref
+
 		parsed, err := reference.Parse(ref)
 		if err != nil {
-			return "", nil, errors.Errorf("invalid oci-layout ref format '%s', must be content-store/image@sha256:digest", ref)
-		}
-		if parsed.Digest() == "" {
-			return "", nil, errors.Errorf("oci-layout ref format '%s' missing digest, must be content-store/image@sha256:digest", ref)
+			return "", nil, err
 		}
-		storeID := parsed.Hostname()
-
-		rslvr = getOCILayoutResolver(storeID, sm, "", g)
+		rslvr = getOCILayoutResolver(parsed.Hostname(), sm, "", g)
```

Comment:
I'm surprised of the use of `Hostname()` in here. Even if you consider this input same as the image reference it doesn't mean that `Hostname()` is the `storeID`. `storeID` is the name that isn't digest or tag in that case. The reason this works is because it uses the containerd reference pkg that only deals with full reference strings(not normalized/familiar). For that library `alpine:latest` is invalid and doesn't work. It needs to be converted to `docker.io/library/alpine` before what doesn't make sense for the OCI case.

So the current code relies on accidental side-effect on passing invalid parameters.

Reply:
Agreed, this is weird. I've reworked the code around this slightly, so that we actually don't do any parsing in `pull.go`. 

I've done the (maybe) controversial thing of passing the not-quite ref `store@digest` (same as we store in the oci identifier) so that the frontend can just pass `named.Name() + "@" + digested.Digest()` and not need to do any fumbling around with dummy names (now all of the dummy ref construction is *explicitly* done in `pull.go`). Since the format is a lot simpler, we don't need to get `.Hostname`, and just take everything before the `@`.